File: crystalos/lib/insight_settings.py
"""Insight Pipeline v2 — settings loader + credit pre-flight.

Two responsibilities:

1. ``load_insight_settings(survey_id, org_id)`` — resolves the 3-level COALESCE
   merge ``survey_insight_settings`` → ``org_insight_defaults`` → platform constant
   defaults (``lib/constants.py``). Tolerates missing tables/rows: on ANY failure it
   returns the platform defaults so the pipeline never crashes on config load.

2. ``credit_preflight(org_id, run_type, settings)`` — resolves the per-run credit
   cost and checks the org balance. **CrystalOS does NOT debit** — the Node backend
   owns the credit ledger and debits on its ``/runs`` path (see module note below).
   CrystalOS reads the balance only for the automated silent-skip decision. For
   manual/refresh/custom runs it raises ``InsufficientCreditsError`` so the caller
   can surface a 402 *before* a run is started (defence-in-depth — the backend has
   already debited, but if CrystalOS is invoked directly we still refuse).

Credit-debit ownership decision (02 §6):
  The credit ledger lives in the Node backend (Postgres credit_accounts /
  credit_ledger, mirrored by backend/src/lib/creditLedger.ts). To avoid a
  double-debit race between two services writing the same ledger, **the backend
  owns debiting**. CrystalOS reads balance for the skip decision only. This is the
  "safe option" called out in the task spec.
"""
from __future__ import annotations

from typing import Any

from crystalos.lib import db
from crystalos.lib.logger import logger
from crystalos.lib import constants as C


# ── Default settings (platform constant floor) ────────────────────────────────

def _platform_defaults() -> dict[str, Any]:
    """Return the platform-constant settings dict (lowest precedence in the merge)."""
    return {
        # Automated incremental
        "automated_insights_enabled":           True,
        "automated_report_generation_enabled":  True,
        "stream_response_threshold":            C.DEFAULT_STREAM_THRESHOLD,
        "report_regen_threshold":               C.DEFAULT_REPORT_REGEN_THRESHOLD,
        "prior_checkpoint_lookback":            C.DEFAULT_PRIOR_CHECKPOINT_LOOKBACK,
        "prior_checkpoint_max_age_days":        C.DEFAULT_PRIOR_CHECKPOINT_MAX_AGE_DAYS,
        "full_checkpoint_response_threshold":   C.DEFAULT_FULL_CHECKPOINT_THRESHOLD,
        "meaningful_delta_nps_points":          C.DEFAULT_MEANINGFUL_DELTA_NPS_POINTS,
        "meaningful_delta_topic_pct":           C.DEFAULT_MEANINGFUL_DELTA_TOPIC_PCT,
        # Refresh
        "refresh_lookback_days":                C.DEFAULT_REFRESH_LOOKBACK_DAYS,
        "refresh_min_response_count":           C.DEFAULT_REFRESH_MIN_RESPONSE_COUNT,
        "refresh_daily_limit":                  C.DEFAULT_REFRESH_DAILY_LIMIT,
        # Manual expert
        "manual_expert_checkpoint_lookback":    C.DEFAULT_MANUAL_EXPERT_CHECKPOINT_LOOKBACK,
        "manual_expert_max_corpus":             C.DEFAULT_MANUAL_EXPERT_MAX_CORPUS,
        "manual_expert_full_corpus_cap":        C.DEFAULT_MANUAL_EXPERT_FULL_CORPUS_CAP,
        "manual_expert_snapshot_count":         C.DEFAULT_MANUAL_EXPERT_SNAPSHOTS,
        # Manual quick
        "manual_quick_sample_cap":              C.DEFAULT_MANUAL_QUICK_SAMPLE,
        "manual_quick_snapshot_count":          C.DEFAULT_MANUAL_QUICK_SNAPSHOTS,
        "manual_quick_default_window_days":     C.DEFAULT_MANUAL_QUICK_WINDOW_DAYS,
        "manual_daily_run_limit":               C.DEFAULT_MANUAL_DAILY_RUN_LIMIT,
        # Custom analysis
        "custom_analysis_enabled":              True,
        "custom_analysis_daily_limit":          3,
        "custom_analysis_max_corpus":           5000,
        "custom_analysis_min_n_for_nps":        30,
        # Credit costs (None in DB ⇒ fall back to these constants)
        "credit_cost_automated_checkpoint":     C.CREDIT_COST_AUTOMATED_CHECKPOINT,
        "credit_cost_automated_report":         C.CREDIT_COST_AUTOMATED_REPORT,
        "credit_cost_refresh":                  C.CREDIT_COST_REFRESH,
        "credit_cost_manual_quick":             C.CREDIT_COST_MANUAL_QUICK,
        "credit_cost_manual_expert":            C.CREDIT_COST_MANUAL_EXPERT,
        # Retention
        "automated_checkpoint_retention_days":  365,
        "manual_report_retention_days":         730,
        "collapse_similar_checkpoints":         True,
        "config_version":                       1,
    }


# Keys carried on org_insight_defaults (subset of survey settings — 03 §13).
_ORG_DEFAULT_KEYS: frozenset[str] = frozenset({
    "automated_insights_enabled", "automated_report_generation_enabled",
    "stream_response_threshold", "prior_checkpoint_lookback",
    "refresh_lookback_days", "refresh_min_response_count", "refresh_daily_limit",
    "manual_daily_run_limit", "manual_expert_checkpoint_lookback",
    "manual_expert_full_corpus_cap", "manual_expert_max_corpus",
    "custom_analysis_enabled", "custom_analysis_daily_limit",
    "credit_cost_automated_checkpoint", "credit_cost_automated_report",
    "credit_cost_refresh", "credit_cost_manual_quick", "credit_cost_manual_expert",
})


async def _fetch_row(table: str, key_col: str, key_val: str) -> dict[str, Any] | None:
    """Fetch a single settings row as a dict, or None if table/row absent. Never raises."""
    try:
        async with db._pool_conn().connection() as conn:
            async with conn.cursor() as cur:
                await cur.execute(
                    f"SELECT * FROM {table} WHERE {key_col} = %s LIMIT 1",  # noqa: S608 - table is a constant literal
                    (key_val,),
                )
                row = await cur.fetchone()
                if row is None:
                    return None
                cols = [d[0] for d in cur.description]
                return dict(zip(cols, row))
    except Exception as exc:
        logger.warning("insight_settings_fetch_failed", table=table, key=key_val, error=str(exc))
        return None


async def load_insight_settings(survey_id: str, org_id: str) -> dict[str, Any]:
    """Resolve effective insight settings via the 3-level COALESCE merge.

    Precedence (highest → lowest):
      survey_insight_settings[survey_id] → org_insight_defaults[org_id] → constants.

    A value in a higher level only wins when it is non-NULL. The result is a plain
    dict keyed by setting name. Always returns a complete dict (every key present)
    even when both tables are missing — falls through to platform defaults.
    """
    merged = _platform_defaults()

    # Level 2: org defaults (only the subset of keys it carries).
    org_row = await _fetch_row("org_insight_defaults", "org_id", org_id)
    if org_row:
        for k in _ORG_DEFAULT_KEYS:
            v = org_row.get(k)
            if v is not None:
                merged[k] = v

    # Level 1: per-survey settings (full key set). Non-NULL wins.
    survey_row = await _fetch_row("survey_insight_settings", "survey_id", survey_id)
    if survey_row:
        for k, v in survey_row.items():
            if k in ("survey_id", "org_id", "updated_at", "updated_by"):
                continue
            if v is not None and k in merged:
                merged[k] = v
            elif v is not None:
                # Carry through any column we did not anticipate (forward-compat).
                merged[k] = v

    return merged


# ── Credit pre-flight ─────────────────────────────────────────────────────────

class InsufficientCreditsError(Exception):
    """Raised by credit_preflight for manual/refresh/custom runs when balance < cost."""

    def __init__(self, required: int, available: int):
        self.required = required
        self.available = available
        super().__init__(f"insufficient_credits: required={required} available={available}")


# Map run_type → settings key + constant fallback for cost resolution.
_COST_KEYS: dict[str, tuple[str, int]] = {
    "automated_incremental": ("credit_cost_automated_checkpoint", C.CREDIT_COST_AUTOMATED_CHECKPOINT),
    "refresh":               ("credit_cost_refresh",              C.CREDIT_COST_REFRESH),
    "manual_quick":          ("credit_cost_manual_quick",         C.CREDIT_COST_MANUAL_QUICK),
    "manual_expert":         ("credit_cost_manual_expert",        C.CREDIT_COST_MANUAL_EXPERT),
    "custom":                ("credit_cost_custom_base",          C.CREDIT_COST_CUSTOM_BASE),
}


def resolve_credit_cost(run_type: str, settings: dict, *, include_report: bool = False) -> int:
    """Resolve the credit cost for a run from settings, falling back to constants.

    For automated runs ``include_report=True`` adds the report-generation surcharge
    (04 §16 routing) on top of the checkpoint cost.
    """
    settings = settings or {}
    key, fallback = _COST_KEYS.get(run_type, _COST_KEYS["automated_incremental"])
    base = settings.get(key)
    try:
        cost = int(base) if base is not None else int(fallback)
    except (TypeError, ValueError):
        cost = int(fallback)
    if run_type == "automated_incremental" and include_report:
        rep = settings.get("credit_cost_automated_report")
        try:
            cost += int(rep) if rep is not None else int(C.CREDIT_COST_AUTOMATED_REPORT)
        except (TypeError, ValueError):
            cost += int(C.CREDIT_COST_AUTOMATED_REPORT)
    return cost


async def get_org_credit_balance(org_id: str) -> int | None:
    """Read the org credit balance from the backend ledger tables.

    Mirrors backend/src/lib/creditLedger.ts: balance = allowance_remaining +
    pack_remaining (overage is unbounded, ignored for the skip check). Returns None
    when the credit tables don't exist (treated as "unlimited / unknown" by callers,
    which then never silently skip). Never raises.
    """
    try:
        async with db._pool_conn().connection() as conn:
            async with conn.cursor() as cur:
                await cur.execute(
                    """SELECT COALESCE(allowance_remaining, 0) + COALESCE(pack_remaining, 0)
                       FROM credit_accounts WHERE org_id = %s LIMIT 1""",
                    (org_id,),
                )
                row = await cur.fetchone()
                if row is None:
                    return None
                return int(row[0] or 0)
    except Exception as exc:
        logger.warning("get_org_credit_balance_failed", org_id=org_id, error=str(exc))
        return None


async def credit_preflight(
    org_id: str,
    run_type: str,
    settings: dict,
    *,
    include_report: bool = False,
) -> bool:
    """Pre-flight credit check (02 §6). CrystalOS reads balance; it does NOT debit.

    Returns:
      True  — sufficient credits (or balance unknown ⇒ do not block).
      False — automated run with insufficient credits → silent skip (logged).

    Raises:
      InsufficientCreditsError — manual / refresh / custom run with insufficient
      credits → caller surfaces HTTP 402.
    """
    cost = resolve_credit_cost(run_type, settings, include_report=include_report)
    balance = await get_org_credit_balance(org_id)

    # Balance unknown (ledger tables absent / dev) → never block; backend is the
    # authoritative gate. This keeps dev + tests running without a ledger.
    if balance is None:
        return True

    if balance < cost:
        if run_type == "automated_incremental":
            logger.info("credit_preflight_insufficient_skip", org_id=org_id,
                        run_type=run_type, required=cost, available=balance)
            return False  # silent skip — caller aborts the run, no error
        raise InsufficientCreditsError(required=cost, available=balance)

    # Sufficient. Debiting is owned by the backend (see module docstring) — no debit here.
    return True

File: backend/src/schemas/insightSettings.ts
/**
 * Zod request schemas for Insight Pipeline v2 settings.
 *
 *  - updateInsightSettingsSchema — PATCH /api/insights/:surveyId/settings body
 *  - updateOrgInsightDefaultsSchema — PATCH /api/orgs/:orgId/insight-defaults body
 *
 * Both are partial (any subset of known keys). Unknown keys are stripped (Zod default).
 * Ranges mirror docs/insights/new_design/03_DATA_MODEL.md §2/§13 and 05_CONFIGURATION.md
 * §3/§7/§10. Credit-cost overrides are positive ints in [1, 500] (platform ceiling) and
 * accept null to mean "fall back to org default / platform constant".
 */
import { z } from 'zod';

// Integer in an inclusive range.
const intRange = (min: number, max: number) =>
  z.number().int().min(min).max(max);

// NUMERIC setting with one decimal of precision, inclusive range.
const numRange = (min: number, max: number) =>
  z.number().min(min).max(max);

// Credit-cost override: positive int 1–500, or null to clear (use lower layer).
const creditCost = z.union([intRange(1, 500), z.null()]);

/**
 * Full per-setting field map. Survey settings accept the entire set; org defaults
 * accept the subset that exists on org_insight_defaults (03 §13).
 */
const settingFields = {
  // Automated incremental
  automated_insights_enabled:           z.boolean(),
  automated_report_generation_enabled:  z.boolean(),
  stream_response_threshold:            intRange(5, 500),
  report_regen_threshold:               intRange(10, 200),
  prior_checkpoint_lookback:            intRange(1, 20),
  prior_checkpoint_max_age_days:        intRange(7, 365),
  full_checkpoint_response_threshold:   intRange(50, 2000),
  meaningful_delta_nps_points:          numRange(0.5, 10),
  meaningful_delta_topic_pct:           numRange(5, 25),

  // Refresh
  refresh_lookback_days:                intRange(7, 365),
  refresh_min_response_count:           intRange(5, 100),
  refresh_daily_limit:                  intRange(1, 20),

  // Manual
  manual_expert_checkpoint_lookback:    intRange(1, 10),
  manual_expert_max_corpus:             intRange(500, 5000),
  manual_expert_full_corpus_cap:        intRange(100, 2000),
  manual_expert_snapshot_count:         intRange(2, 10),
  manual_quick_sample_cap:              intRange(50, 500),
  manual_quick_snapshot_count:          intRange(2, 10),
  manual_quick_default_window_days:     intRange(7, 90),
  manual_daily_run_limit:               intRange(1, 50),

  // Custom Analysis
  custom_analysis_enabled:              z.boolean(),
  custom_analysis_daily_limit:          intRange(1, 20),
  custom_analysis_max_corpus:           intRange(500, 20000),
  custom_analysis_min_n_for_nps:        intRange(10, 100),

  // Credits
  credit_cost_automated_checkpoint:     creditCost,
  credit_cost_automated_report:         creditCost,
  credit_cost_refresh:                  creditCost,
  credit_cost_manual_quick:             creditCost,
  credit_cost_manual_expert:            creditCost,

  // Retention
  automated_checkpoint_retention_days:  intRange(1, 3650),
  manual_report_retention_days:         intRange(1, 3650),
  collapse_similar_checkpoints:         z.boolean(),
} as const;

/**
 * Survey-level settings patch. `.partial()` makes every field optional;
 * `.strict()` rejects unknown keys with a 400 (prevents silent typos).
 */
export const updateInsightSettingsSchema = z
  .object(settingFields)
  .partial()
  .strict()
  .superRefine((val, ctx) => {
    // 05_CONFIGURATION.md §10: manual_expert_max_corpus >= manual_expert_full_corpus_cap.
    // Only enforced when both are present in the same patch (each layer validated alone).
    if (
      val.manual_expert_max_corpus != null &&
      val.manual_expert_full_corpus_cap != null &&
      val.manual_expert_max_corpus < val.manual_expert_full_corpus_cap
    ) {
      ctx.addIssue({
        code: z.ZodIssueCode.custom,
        path: ['manual_expert_max_corpus'],
        message: 'manual_expert_max_corpus must be >= manual_expert_full_corpus_cap',
      });
    }
  });

/**
 * Subset of keys persisted on org_insight_defaults (03 §13). Org defaults are all
 * nullable — a null clears the org-level override so surveys fall through to the
 * platform constant.
 */
const ORG_DEFAULT_KEYS = [
  'automated_insights_enabled',
  'automated_report_generation_enabled',
  'stream_response_threshold',
  'prior_checkpoint_lookback',
  'refresh_lookback_days',
  'refresh_min_response_count',
  'refresh_daily_limit',
  'manual_daily_run_limit',
  'manual_expert_checkpoint_lookback',
  'manual_expert_full_corpus_cap',
  'manual_expert_max_corpus',
  'custom_analysis_enabled',
  'custom_analysis_daily_limit',
  'credit_cost_automated_checkpoint',
  'credit_cost_automated_report',
  'credit_cost_refresh',
  'credit_cost_manual_quick',
  'credit_cost_manual_expert',
] as const;

const orgDefaultFields = Object.fromEntries(
  ORG_DEFAULT_KEYS.map((k) => {
    const base = settingFields[k as keyof typeof settingFields];
    // Booleans / ints become nullable so an org admin can clear an override.
    return [k, base.nullable()];
  }),
) as { [K in (typeof ORG_DEFAULT_KEYS)[number]]: z.ZodNullable<(typeof settingFields)[K]> };

export const updateOrgInsightDefaultsSchema = z
  .object(orgDefaultFields)
  .partial()
  .strict();

export type UpdateInsightSettingsInput = z.infer<typeof updateInsightSettingsSchema>;
export type UpdateOrgInsightDefaultsInput = z.infer<typeof updateOrgInsightDefaultsSchema>;

/** Setting keys persisted on org_insight_defaults (used by the org-defaults route). */
export const ORG_INSIGHT_DEFAULT_KEYS = ORG_DEFAULT_KEYS;
